### PR TITLE
tg3: Common voting interfaces for tgrade

### LINKED ISSUE #85: Use specilaized tg3 version of voting API for tgrade contracts
Reason is to align API to use `points` instead of `weights` in tgrade APIs, but to no affect non-tgrade contracts. It is follow up of: https://github.com/confio/tgrade-contracts/issues/272.

### PR DESCRIPTION
**Pull request by hashedone**
Closes #85

### CODE REVIEW

File: contracts/tgrade-community-pool/src/msg.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@ use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 
 use cosmwasm_std::Coin;
-use cw3::Vote;
+use tg3::Vote;
```
Comment (ethanfrey) [inline contracts/tgrade-community-pool/src/msg.rs:5]: Something seems odd to me here.

If these have different field names, I would expect some diff in this repo when I construct Vote to use `points` not `weight`.

File: packages/voting-contract/src/lib.rs [hunk 2]
```diff
@@ -413,9 +413,9 @@ pub fn list_votes_by_voter(
 pub fn query_voter(deps: Deps, voter: String) -> StdResult<VoterResponse> {
     let cfg = CONFIG.load(deps.storage)?;
     let voter_addr = deps.api.addr_validate(&voter)?;
-    let weight = cfg.group_contract.is_member(&deps.querier, &voter_addr)?;
+    let points = cfg.group_contract.is_member(&deps.querier, &voter_addr)?;
 
-    Ok(VoterResponse { weight })
+    Ok(VoterResponse { points })
```
Comment (ethanfrey) [inline packages/voting-contract/src/lib.rs:418]: Or is this only used in voting-contract and these two updates (here and below) are the only place weight was exposed?

File: packages/voting-contract/src/lib.rs [hunk 3]
```diff
@@ -428,10 +428,7 @@ pub fn list_voters(
         .group_contract
         .list_members(&deps.querier, start_after, limit)?
         .into_iter()
-        .map(|member| VoterDetail {
-            addr: member.addr,
-            weight: member.points,
-        })
+        .map(|Member { addr, points }| VoterDetail { addr, points })
```
Comment (ethanfrey) [inline packages/voting-contract/src/lib.rs:431]: nice restructuring

File: packages/tg3/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+(overriding or just enforcing min/max/default values).
+
+Many implementations will want to restrict who can propose.
+Maybe only people in the voting set. Maybe there is some
+deposit to be made along with the proposal. This is not
+in the spec but left open to the implementation.
+
+`Vote{proposal_id, vote}` - Given a proposal_id, you can
+vote yes, no, abstain or veto. Each signed may have a
+different "weight" in the voting and they apply their
```
Comment (ethanfrey) [inline packages/tg3/README.md:74]: -> points

File: packages/tg3/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+as `Vote`, but for all votes along with pagination.
+Returns the voters sorted by the voters' address in
+lexicographically ascending order.
+
+## Voter Info
+
+Information on who can vote is contract dependent. But
+we will work on a common API to display some of this.
+
+`Voter { address }` - returns voting power (weight) of this address, if any
```
Comment (ethanfrey) [inline packages/tg3/README.md:129]: points

File: packages/tg3/src/query.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+pub struct ProposalListResponse {
+    pub proposals: Vec<ProposalResponse>,
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct VoteListResponse {
+    pub votes: Vec<VoteInfo>,
+}
+
+/// Returns the vote (opinion as well as weight counted) as well as
```
Comment (ethanfrey) [inline packages/tg3/src/query.rs:164]: points (and other usage this file)

File: packages/tg3/src/query.rs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+/// This defines the different ways tallies can happen.
+/// Every contract should support a subset of these, ideally all.
+///
+/// The total_weight used for calculating success as well as the weights of each
+/// individual voter used in tallying should be snapshotted at the beginning of
+/// the block at which the proposal starts (this is likely the responsibility of a
+/// correct cw4 implementation).
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum ThresholdResponse {
```
Comment (ethanfrey) [inline packages/tg3/src/query.rs:19]: Is this object never constructed in the code?
I cannot find anything making "total_points" that would be needed to construct this

**Review by ethanfrey — APPROVED** [review]
I think it is good. Surprised I didn't seem more changes. 
Left a few comments for cleanup, but nothing blocking

File: contracts/tgrade-community-pool/src/msg.rs (hunk 1 above)
Comment (hashedone) [reply contracts/tgrade-community-pool/src/msg.rs:5]: Most of things happens in voting contracts package. In contracts they are just passed around. Thanks to good code abstraction.

File: packages/voting-contract/src/lib.rs (hunk 2 above)
Comment (hashedone) [reply packages/voting-contract/src/lib.rs:418]: Yup.

File: packages/tg3/README.md (hunk 4 above)
Comment (hashedone) [reply packages/tg3/README.md:74]: Ehh, grepped for READMEs and still miss it :/

File: packages/tg3/src/query.rs [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+use cosmwasm_std::{CosmosMsg, Decimal};
+use tg_utils::Expiration;
+
+use crate::msg::Vote;
+
+/// This defines the different ways tallies can happen.
+/// Every contract should support a subset of these, ideally all.
+///
+/// The total_weight used for calculating success as well as the weights of each
```
[the remark this answers is not in the document]
Comment (hashedone) [reply packages/tg3/src/query.rs:13]: I guess I grepped docs and comments for this before moving tg3.

File: packages/tg3/src/query.rs (hunk 7 above)
Comment (hashedone) [reply packages/tg3/src/query.rs:19]: I am surprised - I knew it is in ProposalResponse which is returned by voting-contract, but it turns out, that... ProposalResponse is redefined in voting-contract! Without this one. Just will remove along with the whole ProposalResponse.

**Pull request merged**